pwiz_tools/Skyline/Model/Koina/Models/KoinaModel.cs

```diff
@@ -662,69 +662,77 @@ public static bool KoinaSettingsValid
         /// <param name="mzTolerance">Tolerance for considering two mzs the same</param>
         public static double CalculateSpectrumDotpMzMatch(LibraryRankedSpectrumInfo spectrum1, LibraryRankedSpectrumInfo spectrum2, MzTolerance mzTolerance)
         {
-            var matched1 = spectrum1.PeaksMatched.ToArray();
-            var matched2 = spectrum2.PeaksMatched.ToArray();
-            var intensities1All = new List<double>(matched1.Length + matched2.Length);
-            var intensities2All = new List<double>(matched1.Length + matched2.Length);
+            return CalculateSpectrumDotpMzFull(spectrum1.PeaksMatched, spectrum2.PeaksMatched, mzTolerance, true,
+                true);
+        }
+
+        /// <summary>
+        /// Calculates the normalized contrast angle of the square rooted
+        /// intensities of the two given spectra using all peaks and possibly ignoring
+        /// mismatched peaks
+        /// </summary>
+        /// <param name="spectrum1">First spectrum</param>
+        /// <param name="spectrum2">Second spectrum</param>
+        /// <param name="mzTolerance">Tolerance for considering two mzs the same</param>
+        /// <param name="ignoreMismatches1">if true, unmatched peaks from the first spectrum are ignored</param>
+        /// <param name="ignoreMismatches2">if true, unmatched peaks from the second spectrum are ignored</param>
+        public static double CalculateSpectrumDotpMzFull(IEnumerable<LibraryRankedSpectrumInfo.RankedMI> spectrum1,
+            IEnumerable<LibraryRankedSpectrumInfo.RankedMI> spectrum2, MzTolerance mzTolerance, bool ignoreMismatches1, bool ignoreMismatches2)
+        {
+
+            var s1 = spectrum1.ToArray();
+            var s2 = spectrum2.ToArray();
+            var vectorLength = s1.Length + s2.Length;
+
+            var intensities1All = new List<double>(vectorLength);
+            var intensities2All = new List<double>(vectorLength);
             var matchIndex1 = 0;
             var matchIndex2 = 0;
-            while (matchIndex1 < matched1.Length && matchIndex2 < matched2.Length)
+            while (matchIndex1 < s1.Length && matchIndex2 < s2.Length)
             {
-                var mz1 = matched1[matchIndex1].ObservedMz;
-                var mz2 = matched2[matchIndex2].ObservedMz;
+                var mz1 = s1[matchIndex1].ObservedMz;
+                var mz2 = s2[matchIndex2].ObservedMz;
                 if (mzTolerance.IsWithinTolerance(mz1, mz2))
                 {
-                    intensities1All.Add(matched1[matchIndex1].Intensity);
-                    intensities2All.Add(matched2[matchIndex2].Intensity);
-
-                    ++matchIndex1;
-                    ++matchIndex2;
+                    intensities1All.Add(s1[matchIndex1++].Intensity);
+                    intensities2All.Add(s2[matchIndex2++].Intensity);
                 }
-                else if (mz1 < mz2)
+
+                if (mz1 < mz2)

```

Comment:
I am pretty sure this needs to be "else if (mz1 < mz2)" because otherwise "matchIndex1" or "matchIndex2" will potentially get incremented twice.